torch/_dynamo/variables/torch_function.py

```diff
@@ -0,0 +1,195 @@
+        self.subclass_type = subclass_type
+
+    def as_proxy(self):
+        return self.tensor_variable.as_proxy()
+
+    def python_type(self):
+        return self.subclass_type
+
+    def torch_function_var(self, tx):
+        from .builder import VariableBuilder
+
+        source = AttrSource(
+            AttrSource(self.source, "__torch_function__"),
+            "__func__",
+        )
+        return VariableBuilder(tx, source)(self.torch_function_fn)
+
+    def subclass_type_var(self):
+        return UserDefinedClassVariable(self.subclass_type)
+
+    def global_class_name(self):
+        return f"__subclass_{self.subclass_type.__name__}"
+
+    def call_torch_function(self, tx, fn, types, args, kwargs):
+        return call_torch_function(
+            tx,
+            self.subclass_type_var(),
+            build_torch_function_var(tx, self.torch_function_fn, self.source),
+            fn,
+            types,
+            args,
+            kwargs,
+        )
+
+    def call_method(
+        self,
+        tx,
+        name,
+        args: "List[VariableTracker]",
+        kwargs: "Dict[str, VariableTracker]",
+    ) -> "VariableTracker":
+        # This code block implements inlining the __torch_function__ override
+        # of `call_method`.
+        if tx.output.torch_function_enabled:
+            import torch
+            from .builder import SourcelessBuilder
+
+            options = VariableTracker.propagate(self, args, kwargs.values())
+            args = [self] + list(args)
+            # [Note: __torch_function__] Currently we only support methods that are defined on tensor
+            # we will graph break in other cases this will need a bigger overhaul of extracting methods/comparing them for equality
+            func_var = SourcelessBuilder()(tx, getattr(torch.Tensor, name))
+            all_args = args + tree_flatten(kwargs)[0]
+            types = TupleVariable(
+                [
+                    a.subclass_type_var()
+                    for a in all_args
+                    if isinstance(a, TensorWithTFOverrideVariable)
+                ]
+            )
+
+            return self.call_torch_function(tx, func_var, types, args, kwargs)
+        else:
+            return self.tensor_variable.call_method(tx, name, args, kwargs)
+
+
+def can_dispatch_torch_function(tx, args, kwargs):
+    if tx.output.torch_function_enabled:
+        all_args = tree_flatten(args)[0] + tree_flatten(kwargs)[0]
+        return any(isinstance(arg, TensorWithTFOverrideVariable) for arg in all_args)
+    else:
+        return False
+
+
+def dispatch_torch_function(tx, fn, args, kwargs):
+    """Gathers all args that are TensorWithTFOverrideVariable and dispatches based on the ordering in _get_overloaded_args"""
+
+    all_args = args + tree_flatten(kwargs)[0]
+    overloaded_args = _get_overloaded_args(
+        [arg for arg in all_args if isinstance(arg, TensorWithTFOverrideVariable)],
```

Comment:
This is insufficient because many functions are able to dispatch on torch function even if it is in a list of tensors

Reply:
And it looks asymmetric with `can_dispatch_torch_function` which flattens args

Reply:
Yeah I need to flatten here, I'll add a test for this too